### ISSUE TITLE
[BUG][PyTorch 2.0 Export][quant]:get_source_partitions() may return different matches with same input graph

### ISSUE THREAD

**Issue by GodHforever** [pytorch/pytorch#147170]
### 🐛 Describe the bug

I am attempting to extend the quantization backend based on PyTorch 2.0 export. The operator I have chosen is `torch.gather` . 
The input code I tested is as follows
```python
    class GatherLayer(nn.Module):
        def forward(self, x):
            assert x.shape == (2,2)
            x = torch.gather(x, dim=0, index=torch.tensor([[0, 0], [1, 0]]))
            return x

    example_inputs = (torch.tensor([[0,1],[0,1]]),)

    model = GatherLayer()
    model.eval()
    exported_model = torch.export.export(model, example_inputs).module()
    # print(exported_model.graph)

    prepared_model = prepare_pt2e(exported_model, quantizer)
    prepared_model(*example_inputs)
    # print(prepared_model.graph)
    quantized_model = convert_pt2e(prepared_model)
```

. The part of the code I constructed for the quantization backend is as follows: 
```python
def _annotate_gather(
    gm: torch.fx.GraphModule,
    quantization_config: Optional[QuantizationConfig],
    filter_fn: Optional[Callable[[Node], bool]] = None,
) -> Optional[List[List[Node]]]:
    print(gm.graph)
    partitions = get_source_partitions(gm.graph, [torch.gather], filter_fn)
    matches = list(itertools.chain.from_iterable(partitions.values()))
    annotated_partitions = []

    for matche in matches:
        output_nodes = matche.output_nodes
        input_nodes = matche.input_nodes
        gather_node = output_nodes[0]
        input_qspec_map = {}
        partition = []
        input_node = input_nodes[1] # sometimes input_nodes[1] is input, while sometimes input_nodes[0] is input 
        input_qspec_map[input_node] = get_input_act_qspec(quantization_config)
        partition.append(input_node)

        gather_node.meta["quantization_annotation"] = QuantizationAnnotation(
            input_qspec_map=input_qspec_map,
            output_qspec=get_output_act_qspec(quantization_config),
            _annotated=True,
        )
        _mark_nodes_as_annotated(partition)
        annotated_partitions.append(partition)
    return annotated_partitions
```

 I found that with the same input, the matching results can differ, **_specifically in the order of `input_nodes`_** . Input graph is:
```
graph():
    %lifted_tensor_0 : [num_users=1] = get_attr[target=lifted_tensor_0]
    %x : [num_users=1] = placeholder[target=x]
    %lift_fresh_copy : [num_users=1] = call_function[target=torch.ops.aten.lift_fresh_copy.default](args = (%lifted_tensor_0,), kwargs = {})
    %detach : [num_users=1] = call_function[target=torch.ops.aten.detach.default](args = (%lift_fresh_copy,), kwargs = {})
    %gather : [num_users=1] = call_function[target=torch.ops.aten.gather.default](args = (%x, 0, %detach), kwargs = {})
    return (gather,)
```
 after matching with  `get_source_partitions` , the order of  node `x` and node `detach` in `input_nodes`  is not same all the time, which causes errors in my subsequent code.
**I totally run 100 times, and In 25 of these cases node `x` appears in `input_node[1]` and the rest in `input_node[0]`** 

Notice:
I tried the `SubgraphMatcherWithNameNodeMap` , but the pattern of `torch.gather` is not easy to describe due to 'index`, so I turned to this api.

cc @chauhang @penguinwu @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

### Versions

Collecting environment information...
PyTorch version: 2.5.1+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.28.4
Libc version: glibc-2.31

Python version: 3.11.11 (main, Dec 11 2024, 16:28:39) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-200-generic-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 48 bits virtual
CPU(s):                             96
On-line CPU(s) list:                0-95
Thread(s) per core:                 2
Core(s) per socket:                 24
Socket(s):                          2
NUMA node(s):                       2
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              85
Model name:                         Intel(R) Xeon(R) Platinum 8369HB CPU @ 3.30GHz
Stepping:                           11
CPU MHz:                            3800.047
CPU max MHz:                        4200.0000
CPU min MHz:                        1200.0000
BogoMIPS:                           6600.06
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          1.5 MiB
L1i cache:                          1.5 MiB
L2 cache:                           48 MiB
L3 cache:                           66 MiB
NUMA node0 CPU(s):                  0-47
NUMA node1 CPU(s):                  48-95
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        KVM: Vulnerable
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:             Vulnerable
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Retpoline
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 ida arat avx512_vnni

Versions of relevant libraries:
[pip3] intel_extension_for_pytorch==2.5.0
[pip3] numpy==1.26.3
[pip3] onnx==1.17.0
[pip3] torch==2.5.1+cpu
[pip3] torchaudio==2.5.1+cpu
[pip3] torchvision==0.20.1+cpu
[conda] intel-extension-for-pytorch 2.5.0                    pypi_0    pypi
[conda] mkl-include               2025.0.1                 pypi_0    pypi
[conda] mkl-static                2025.0.1                 pypi_0    pypi
[conda] numpy                     1.26.3                   pypi_0    pypi
[conda] torch                     2.5.1+cpu                pypi_0    pypi
[conda] torchaudio                2.5.1+cpu                pypi_0    pypi
[conda] torchvision               0.20.1+cpu               pypi_0    pypi

**Comment by urstrulyvishtan** [comment]
If this behavior is confirmed as an issue with `get_source_partitions`,  we can request feature to ensure deterministic behavior in such cases.

**Comment by angelayi** [comment]
In the case where `input_nodes[1]` is `x`, what is the value of `input_nodes[0]`?

**Comment by GodHforever** [comment]
> In the case where `input_nodes[1]` is `x`, what is the value of `input_nodes[0]`?

`input_nodes[0]` is `detach`。They are just switched the order.

**Comment by GodHforever** [comment]
> If this behavior is confirmed as an issue with `get_source_partitions`, we can request feature to ensure deterministic behavior in such cases.
> 
> Are there any specific conditions under which the order of `input_nodes` changes? For instance, does it depend on the input graph structure or specific attributes of the nodes?

In my tests, the whole process is identical, nothing changes, but the node order changes. I'm not sure if there is  any specific struceture or attributes in my test case that triggers this issue.

**Comment by angelayi** [comment]
This is because we use a [set](https://github.com/pytorch/pytorch/blob/09291817b29a70fa20d97207fcfedd0d1dc6759b/torch/fx/passes/utils/source_matcher_utils.py#L102) instead of something like an ordered set(?). I wouldn't say this is a bug though, since we never specified that it should appear in the same order.

**Comment by hgleung** [comment]
> This is because we use a [set](https://github.com/pytorch/pytorch/blob/09291817b29a70fa20d97207fcfedd0d1dc6759b/torch/fx/passes/utils/source_matcher_utils.py#L102) instead of something like an ordered set(?). I wouldn't say this is a bug though, since we never specified that it should appear in the same order.

I'd like to contribute as a first-time contributor, but like you said, it isn't exactly a bug since it isn't specified that it should appear in the same order, so what is the issue to be addressed? Should the input_nodes be changed to an ordered set (or other data structure) that ensures the order is maintained?

**Comment by urstrulyvishtan** [comment]
> This is because we use a [set](https://github.com/pytorch/pytorch/blob/09291817b29a70fa20d97207fcfedd0d1dc6759b/torch/fx/passes/utils/source_matcher_utils.py#L102) instead of something like an ordered set(?). I wouldn't say this is a bug though, since we never specified that it should appear in the same order.

@angelayi 
RFC:
Will sort the list when it's first created, focusing on `input_nodes` always has a consistent order based on a unique property of each node.

Let me know if this makes sense

**Comment by julian-reed** [comment]
It looks like this issue was close to being completed previously, but since all previous PRs have been closed, I will resolve this issue in a new one.

**Comment by lakshmiprasanna07** [comment]
Hi! I think I found the cause of this.

`make_partition()` in `source_matcher_utils.py` builds `input_nodes` (and similarly `output_nodes` and `params`) using plain sets. Since `torch.fx.Node` doesn't define `__eq__` or `__hash__`, the objects are identity-hashed, and the iteration order of a `set` depends on the hash table layout rather than the graph structure itself. That would explain why the partition ordering isn't deterministic across runs.

I verified this by tracing the same graph 200 times and recording the order of `input_nodes`. Both orderings from the issue appeared across those runs.

My proposed fix is to preserve insertion order by collecting these nodes in a `dict` (using `node` as the key) and converting the keys to a `list` at the end, instead of using a `set`. I ran the existing `test_source_matcher_utils.py` suite with this change and all tests still pass.

I'm going to open a PR with this fix along with a regression test that reproduces the original nondeterministic behavior.
